crypto/ube/ube.c

```diff
@@ -0,0 +1,238 @@
+
+  ube_state.generation_number += 1;
+
+  // Make sure we cache all new generation numbers. Otherwise, we might detect
+  // a fork UBE but, in fact, both a fork and snapsafe UBE occurred. Then next
+  // time we enter, a redundant reseed will be emitted.
+  ube_state.cached_fork_gn = current_detection_gn->current_fork_gn;
+  ube_state.cached_snapsafe_gn = current_detection_gn->current_snapsafe_gn;
+}
+
+// ube_get_detection_generation_numbers loads the current detection generation
+// numbers into |current_detection_gn|.
+//
+// Returns 1 on success and 0 otherwise. The 0 return value means that a
+// detection method we expected to be available, is in fact not.
+static int ube_get_detection_generation_numbers(
+  struct detection_gn *current_detection_gn) {
+
+  // An attempt to prevent a situation where new detection methods are added but
+  // we forget to load them.
+  OPENSSL_STATIC_ASSERT(NUMBER_OF_DETECTION_GENERATION_NUMBERS == 2,
+    not_handling_the_exact_number_of_detection_generation_numbers);
+
+  current_detection_gn->current_fork_gn = get_fork_generation_number_mocked();
+  int ret_snapsafe_gn = get_snapsafe_generation_number_mocked(
+                          &(current_detection_gn->current_snapsafe_gn));
+
+  if (current_detection_gn->current_fork_gn == 0 ||
+      ret_snapsafe_gn == 0) {
+    return 0;
+  }
+
+  return 1;
+}
+
+// ube_is_detected computes whether a UBE has been detected or not by comparing
+// the cached detection generation numbers with the current detection generation
+// numbers. The current generation numbers must be loaded into
+// |current_detection_gn| before calling this function.
+//
+// Returns 1 if UBE has been detected and 0 if no UBE has been detected.
+static int ube_is_detected(struct detection_gn *current_detection_gn) {
+
+  if (ube_state.cached_fork_gn != current_detection_gn->current_fork_gn ||
+      ube_state.cached_snapsafe_gn != current_detection_gn->current_snapsafe_gn) {
+    return 1;
+  }
+  return 0;
+}
+
+// stays invariant after a failure state has been reached.
+int get_ube_generation_number(uint64_t *current_generation_number) {
+
+  GUARD_PTR(current_generation_number);
+  *current_generation_number = 0;
+
+  CRYPTO_once(&ube_state_initialize_once, ube_state_initialize);
+
+  // If something failed at an earlier point short-circuit immediately. This
+  // saves work in case the UBE detection is not supported. The check below
+  // must be done after attempting to initialize the UBE state. Because
+  // initialization might fail and we can short-circuit here.
+  if (ube_detection_unavailable == 1) {
+    return 0;
+  }
+
+  struct detection_gn current_detection_gn = { 0 };
+
+  // First read generation numbers for each supported detection method. We do
+  // not mutate |ube_state|. So, a read lock is sufficient at this point. Each
+  // individual detection method will have their own concurrency controls if
+  // needed.
+
+  CRYPTO_STATIC_MUTEX_lock_read(&ube_lock);
```

Comment:
Why lock here? Nothing in ube_get_detection_generation_numbers touches ube_state, I think we only need the read lock around ube_is_detected.

Reply:
Hmm, yeah I think that would work.

Reply:
Obviously, this require `ube_get_detection_generation_numbers` to be thread-safe. But we explicitly do make each detection function thread-safe as long as any shared argument memory is serialised. But no shared argument memory in this case, so we are good.